File: src/main/java/ltdjms/discord/aiagent/commands/AgentCompletionListener.java
package ltdjms.discord.aiagent.commands;

import java.util.List;
import java.util.function.Consumer;
import javax.inject.Inject;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import ltdjms.discord.aichat.services.MessageSplitter;
import ltdjms.discord.shared.di.JDAProvider;
import ltdjms.discord.shared.events.AgentCompletedEvent;
import ltdjms.discord.shared.events.AgentFailedEvent;
import ltdjms.discord.shared.events.DomainEvent;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel;
import net.dv8tion.jda.api.entities.channel.middleman.MessageChannel;

/**
 * Agent 完成監聽器。
 *
 * <p>監聽 {@link AgentCompletedEvent} 和 {@link AgentFailedEvent}， 將最終回應或錯誤訊息發送給 Discord 用戶。
 */
public final class AgentCompletionListener implements Consumer<DomainEvent> {

  private static final Logger LOGGER = LoggerFactory.getLogger(AgentCompletionListener.class);

  @Inject
  public AgentCompletionListener() {
    // Dagger 構造函數
  }

  @Override
  public void accept(DomainEvent event) {
    if (event instanceof AgentCompletedEvent e) {
      handleAgentCompleted(e);
    } else if (event instanceof AgentFailedEvent e) {
      handleAgentFailed(e);
    }
  }

  /**
   * 處理 Agent 完成事件。
   *
   * @param event 完成事件
   */
  private void handleAgentCompleted(AgentCompletedEvent event) {
    try {
      MessageChannel channel = resolveMessageChannel(event.guildId(), event.channelId());
      if (channel == null) {
        LOGGER.warn("無法取得可發送訊息的頻道: guildId={}, channelId={}", event.guildId(), event.channelId());
        return;
      }

      List<String> messages = MessageSplitter.split(event.finalResponse());
      for (String message : messages) {
        channel.sendMessage(message).queue();
      }

      LOGGER.info(
          "Agent 完成，已發送最終回應: conversationId={}, length={}",
          event.conversationId(),
          messages.size());

    } catch (Exception e) {
      LOGGER.error("處理 Agent 完成事件時發生錯誤", e);
    }
  }

  /**
   * 處理 Agent 失敗事件。
   *
   * @param event 失敗事件
   */
  private void handleAgentFailed(AgentFailedEvent event) {
    try {
      MessageChannel channel = resolveMessageChannel(event.guildId(), event.channelId());
      if (channel == null) {
        LOGGER.warn("無法取得可發送訊息的頻道: guildId={}, channelId={}", event.guildId(), event.channelId());
        return;
      }

      String errorMessage = "❌ " + event.reason();
      channel.sendMessage(errorMessage).queue();

      LOGGER.warn(
          "Agent 失敗，已發送錯誤訊息: conversationId={}, reason={}", event.conversationId(), event.reason());

    } catch (Exception e) {
      LOGGER.error("處理 Agent 失敗事件時發生錯誤", e);
    }
  }

  /**
   * 解析訊息頻道。
   *
   * @param guildId 伺服器 ID
   * @param channelId 頻道 ID 字串
   * @return 訊息頻道，如果無法解析則返回 null
   */
  private MessageChannel resolveMessageChannel(long guildId, String channelId) {
    try {
      Guild guild = JDAProvider.getJda().getGuildById(guildId);
      if (guild == null) {
        return null;
      }

      long channelIdLong = Long.parseLong(channelId);
      GuildChannel guildChannel = guild.getGuildChannelById(channelIdLong);
      if (guildChannel instanceof MessageChannel messageChannel) {
        return messageChannel;
      }

      return guild.getThreadChannelById(channelIdLong);
    } catch (NumberFormatException e) {
      LOGGER.warn("無法解析頻道 ID: {}", channelId);
      return null;
    }
  }
}

File: packages/ai/src/listeners/agent-completion-listener.ts
import type { DomainEvent, DiscordRuntimeGateway } from '@ltdjms/shared';
import type { AgentCompletedEvent, AgentFailedEvent } from '@ltdjms/shared';
import type { GuildTextBasedChannel } from 'discord.js';
import { MessageSplitter } from '../services/MessageSplitter.js';
import { DiscordMarkdownStreamProcessor } from '../markdown/services/DiscordMarkdownStreamProcessor.js';
import { MarkdownHeadingSegmenter } from '../markdown/services/MarkdownHeadingSegmenter.js';
import type { CommonMarkValidator } from '../markdown/validation/CommonMarkValidator.js';
import type { MarkdownAutoFixer } from '../markdown/autofix/MarkdownAutoFixer.js';
import type { DiscordMarkdownSanitizer } from '../markdown/services/DiscordMarkdownSanitizer.js';
import type { DiscordMarkdownPaginator } from '../markdown/services/DiscordMarkdownPaginator.js';
import pino from 'pino';

const EMPTY_RESPONSE_FALLBACK = ':question: AI 沒有產生回應';

export interface AgentCompletionMarkdownPipeline {
  validator: CommonMarkValidator;
  autoFixer: MarkdownAutoFixer;
  sanitizer: DiscordMarkdownSanitizer;
  paginator: DiscordMarkdownPaginator;
}

/**
 * Sends Discord notifications when agent sessions complete or fail.
 * Matches Java AgentCompletionListener.
 */
export class AgentCompletionListener {
  private readonly splitter = new MessageSplitter();
  private readonly logger: pino.Logger;

  constructor(
    private readonly runtimeGateway: DiscordRuntimeGateway,
    logger?: pino.Logger,
    private readonly markdownPipeline?: AgentCompletionMarkdownPipeline,
  ) {
    this.logger = logger ?? pino({ name: 'agent-completion-listener' });
  }

  accept(event: DomainEvent | null | undefined): void {
    if (!event) {
      return;
    }

    if (event.eventType === 'agent_completed') {
      void this.handleAgentCompleted(event as AgentCompletedEvent).catch((error) => {
        this.logger.error({ err: error }, '處理 Agent 完成事件時發生錯誤');
      });
      return;
    }

    if (event.eventType === 'agent_failed') {
      this.handleAgentFailed(event as AgentFailedEvent);
    }
  }

  private async handleAgentCompleted(event: AgentCompletedEvent): Promise<void> {
    const channel = this.resolveMessageChannel(event.guildId, event.channelId);
    if (!channel) {
      this.logger.warn(
        { guildId: event.guildId, channelId: event.channelId },
        '無法取得可發送訊息的頻道',
      );
      return;
    }

    const pages = this.prepareFinalPages(event.finalResponse);
    let sentAny = false;
    for (const message of pages) {
      if (!message?.trim()) {
        continue;
      }
      await channel.send(message);
      sentAny = true;
    }

    if (!sentAny) {
      await channel.send(EMPTY_RESPONSE_FALLBACK);
    }

    this.logger.info(
      { conversationId: event.conversationId, parts: pages.length },
      'Agent 完成，已發送最終回應',
    );
  }

  private handleAgentFailed(event: AgentFailedEvent): void {
    // Errors are surfaced by AIChatMentionListener via streaming handler.onChunk(error).
    // Publishing agent_failed remains for observability; avoid duplicate Discord messages.
    this.logger.warn(
      { conversationId: event.conversationId, reason: event.reason },
      'Agent 失敗（已由 mention listener 通知使用者）',
    );
  }

  private prepareFinalPages(finalResponse: string): string[] {
    if (!this.markdownPipeline) {
      return this.splitter.split(finalResponse);
    }

    const processor = new DiscordMarkdownStreamProcessor(
      new MarkdownHeadingSegmenter(),
      this.markdownPipeline.validator,
      this.markdownPipeline.autoFixer,
      this.markdownPipeline.sanitizer,
      this.markdownPipeline.paginator,
    );
    const pages = [...processor.onChunk(finalResponse), ...processor.flush()];
    if (pages.length === 0) {
      return this.splitter.split(finalResponse);
    }
    return pages;
  }

  private resolveMessageChannel(guildId: string, channelId: string): GuildTextBasedChannel | null {
    if (!/^\d+$/.test(channelId)) {
      this.logger.warn({ channelId }, '無法解析頻道 ID');
      return null;
    }

    const channel =
      this.runtimeGateway.findGuildChannel(guildId, channelId) ??
      this.runtimeGateway.findThreadChannel(guildId, channelId);

    if (!channel || typeof channel !== 'object') {
      return null;
    }

    const candidate = channel as { isTextBased?: () => boolean };
    if (typeof candidate.isTextBased === 'function' && candidate.isTextBased()) {
      return channel as GuildTextBasedChannel;
    }

    return null;
  }
}
